File: app/src/main/java/com/shevelev/comics_viewer/dal/OptionsDal.java
package com.shevelev.comics_viewer.dal;

import android.util.Log;

import com.activeandroid.ActiveAndroid;
import com.activeandroid.query.Delete;
import com.activeandroid.query.Select;
import com.shevelev.comics_viewer.common.helpers.CollectionsHelper;
import com.shevelev.comics_viewer.dal.dto.Option;
import com.shevelev.comics_viewer.dal.entities.DbOption;

import java.util.ArrayList;
import java.util.List;

/**
 * Dal for innerOptionsCollection
 */
public class OptionsDal implements IOptionsDal
{
    /**
     * @return false - fail
     */
    private void add(Option option)
    {
        new DbOption(option).save();
    }

    /**
     * @return false - fail
     */
    private void update(Option option)
    {
        ArrayList<DbOption> dbResult = new Select().from(DbOption.class).where("Key = ?", option.key).execute();

        if (dbResult.size() == 0)
            return;

        DbOption dbOption = dbResult.get(0);
        dbOption.value = option.value;
        dbOption.save();
    }

    public boolean update(List<Option> optionsToAdd, List<Option> optionsToUpdate)
    {
        ActiveAndroid.beginTransaction();
        try
        {
            for(Option option : optionsToAdd)
                add(option);

            for(Option option : optionsToUpdate)
                update(option);

            ActiveAndroid.setTransactionSuccessful();
            return true;
        }
        catch (Exception ex)
        {
            Log.e("CV", "exception", ex);
            return false;
        }
        finally
        {
            ActiveAndroid.endTransaction();
        }
    }

    /**
     * @return false - fail
     */
    @Override
    public boolean delete(int[] keys)
    {
        ActiveAndroid.beginTransaction();
        try
        {
            for(int key : keys)
                new Delete().from(DbOption.class).where("Key = ?", key).execute();

            return true;
        }
        catch (Exception ex)
        {
            Log.e("CV", "exception", ex);
            return false;
        }
        finally
        {
            ActiveAndroid.endTransaction();
        }
    }

    /**
     * @return null - fail
     */
    @Override
    public List<Option> getAll()
    {
        try
        {
            ArrayList<DbOption> dbResult=new Select().from(DbOption.class).execute();
            return CollectionsHelper.transform(dbResult, item -> new Option(item));
        }
        catch (Exception ex)
        {
            Log.e("CV", "exception", ex);
            return null;
        }
    }
}

File: app/src/main/java/com/shevelev/comics_viewer/dal/OptionsDal.kt
package com.shevelev.comics_viewer.dal

import android.util.Log
import com.activeandroid.ActiveAndroid
import com.activeandroid.Model
import com.activeandroid.query.Delete
import com.activeandroid.query.Select
import com.shevelev.comics_viewer.common.helpers.CollectionsHelper
import com.shevelev.comics_viewer.dal.dto.Option
import com.shevelev.comics_viewer.dal.entities.DbOption

/**
 * Dal for innerOptionsCollection
 */
class OptionsDal : IOptionsDal {
    /**
     * @return false - fail
     */
    private fun add(option: Option?) {
        DbOption(option!!).save()
    }

    /**
     * @return false - fail
     */
    private fun update(option: Option?) {
        val dbResult = Select().from(DbOption::class.java).where("Key = ?", option!!.key).execute<DbOption>()
        if (dbResult.size == 0) return
        val dbOption = dbResult[0]
        dbOption.value = option.value
        dbOption.save()
    }

    override fun update(optionsToAdd: List<Option?>?, optionsToUpdate: List<Option?>?): Boolean {
        ActiveAndroid.beginTransaction()
        return try {
            for (option in optionsToAdd!!) add(option)
            for (option in optionsToUpdate!!) update(option)
            ActiveAndroid.setTransactionSuccessful()
            true
        } catch (ex: Exception) {
            Log.e("CV", "exception", ex)
            false
        } finally {
            ActiveAndroid.endTransaction()
        }
    }

    /**
     * @return false - fail
     */
    override fun delete(keys: IntArray?): Boolean {
        ActiveAndroid.beginTransaction()
        return try {
            for (key in keys!!) Delete().from(DbOption::class.java).where("Key = ?", key).execute<Model>()
            true
        } catch (ex: Exception) {
            Log.e("CV", "exception", ex)
            false
        } finally {
            ActiveAndroid.endTransaction()
        }
    }

    /**
     * @return null - fail
     */
    override val all: List<Option>?
        get() = try {
            val dbResult = Select().from(DbOption::class.java).execute<DbOption>()
            CollectionsHelper.transform(dbResult) { item: DbOption? -> Option(item!!) }
        } catch (ex: Exception) {
            Log.e("CV", "exception", ex)
            null
        }
}